server/automute_channel.go

```diff
@@ -0,0 +1,94 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/mattermost/mattermost/server/public/plugin"
+
+	"github.com/pkg/errors"
+)
+
+func (p *Plugin) UserHasJoinedChannel(c *plugin.Context, channelMember *model.ChannelMember, actor *model.User) {
+	_, _ = p.updateAutomutingOnUserJoinedChannel(c, channelMember.UserId, channelMember.ChannelId)
+}
+
+func (p *Plugin) updateAutomutingOnUserJoinedChannel(c *plugin.Context, userID string, channelID string) (bool, error) {
+	if automuteEnabled := p.getAutomuteIsEnabledForUser(userID); !automuteEnabled {
+		return false, nil
+	}
+
+	if canAutomute, err := p.isChannelLinked(channelID); err != nil {
+		p.API.LogError(
+			"Unable to check if channel is linked to update automuting when a user has joined the channel",
+			"UserID", userID,
+			"ChannelID", channelID,
+			"Error", err.Error(),
+		)
+		return false, errors.Wrap(err, "Unable to update automuting when a user has joined a channel")
+	} else if !canAutomute {
+		// Only automute channels that are linked
+		return false, nil
+	}
+
+	err := p.setChannelMembersAutomuted([]*model.ChannelMemberIdentifier{{UserId: userID, ChannelId: channelID}}, true)
+	return err == nil, err
+}
+
+func (p *Plugin) ChannelHasBeenCreated(c *plugin.Context, channel *model.Channel) {
+	_ = p.updateAutomutingOnChannelCreated(channel)
+}
+
+func (p *Plugin) updateAutomutingOnChannelCreated(channel *model.Channel) error {
+	if !channel.IsGroupOrDirect() {
+		// Assume that newly created channels can never be linked by the time this is called
+		return nil
+	}
+
+	return p.updateAutomutingForChannelMembers(channel.Id, true)
+}
+
+func (p *Plugin) updateAutomutingOnChannelLinked(channelID string) error {
+	// This simply mutes the channel for all users with automuting enabled, regardless of their settings before. It
+	// doesn't pay attention to if the user manually muted the channel beforehand.
+	return p.updateAutomutingForChannelMembers(channelID, true)
+}
+
+func (p *Plugin) updateAutomutingOnChannelUnlinked(channelID string) error {
+	// This simply unmutes the channel for all users with automuting enabled, regardless of their settings before. It
+	// doesn't pay attention to if the user manually muted the channel beforehand to keep it muted.
+	return p.updateAutomutingForChannelMembers(channelID, false)
+}
+
+func (p *Plugin) updateAutomutingForChannelMembers(channelID string, enableAutomute bool) error {
+	var membersToUpdate []*model.ChannelMemberIdentifier
+
+	page := 0
+	perPage := 200
+	for {
```

Comment:
nit: inline?
```
for page := 0; ; page += 1 {
    // ...
}
```

Reply:
In retrospect, it really feels like we should've done this paging on the server-side and let plugins just get an unpaginated list. I can't think of any cases where plugin code wouldn't want to get all the objects at once, so it seems like we're going to see this pattern a lot

Reply:
I do think it's valuable to be able to request less than the whole payload, otherwise we'd run into potential bottlenecks with large requests. We can always abstract away the pagination in the pluginapi at some point :)

Reply:
I'd rather leave this one as-is since it can't be written using the middle condition parameter without jumping through some hoops with the scope of `members`